### PR TITLE
Add Unicode block-drawing compiler output support

### PR DESCRIPTION
**Pull request by estebank**
Add nightly-only theming support to rustc output using Unicode box
drawing characters instead of ASCII-art to draw the terminal UI.

In order to enable, the flags `-Zunstable-options=yes --error-format=human-unicode` must be passed in.

After:

```
error: foo
  ╭▸ test.rs:3:3
  │
3 │       X0 Y0 Z0
  │   ┌───╿──│──┘
  │  ┌│───│──┘
  │ ┏││━━━┙
  │ ┃││
4 │ ┃││   X1 Y1 Z1
5 │ ┃││   X2 Y2 Z2
  │ ┃│└────╿──│──┘ `Z` label
  │ ┃└─────│──┤
  │ ┗━━━━━━┥  `Y` is a good letter too
  │        `X` is a good letter
  ╰╴
note: bar
  ╭▸ test.rs:4:3
  │
4 │ ┏   X1 Y1 Z1
5 │ ┃   X2 Y2 Z2
6 │ ┃   X3 Y3 Z3
  │ ┗━━━━━━━━━━┛
  ├ note: bar
  ╰ note: baz
note: qux
  ╭▸ test.rs:4:3
  │
4 │   X1 Y1 Z1
  ╰╴  ━━━━━━━━
```

Before:

```
error: foo
 --> test.rs:3:3
  |
3 |       X0 Y0 Z0
  |    ___^__-__-
  |   |___|__|
  |  ||___|
  | |||
4 | |||   X1 Y1 Z1
5 | |||   X2 Y2 Z2
  | |||____^__-__- `Z` label
  | ||_____|__|
  | |______|  `Y` is a good letter too
  |        `X` is a good letter
  |
note: bar
 --> test.rs:4:3
  |
4 | /   X1 Y1 Z1
5 | |   X2 Y2 Z2
6 | |   X3 Y3 Z3
  | |__________^
  = note: bar
  = note: baz
note: qux
 --> test.rs:4:3
  |
4 |   X1 Y1 Z1
  |   ^^^^^^^^
```

After:

rustc output with unicode box drawing characters

Before:
current rustc output with ASCII art

### CODE REVIEW

**Comment by estebank** [comment]
r? compiler

File: compiler/rustc_errors/src/emitter.rs
Comment (estebank) [inline compiler/rustc_errors/src/emitter.rs:1387]: The commit that introduces the `|` to multline `note`s doesn't need to land with this PR at all. I can even leave the ASCII art as is and only put the vertical line in the unicode output exclusively.

File: tests/ui/check-cfg/cargo-build-script.stderr
Comment (Urgau) [inline tests/ui/check-cfg/cargo-build-script.stderr:11]: it's a bit unfortunate that it breaks the simple copy/pasting of this suggestion[^1], but you can see that the suggestion is currently off-by-one because there wasn't any vertical line, so it's a loss for a win 

[^1]: suggestion which isn't (for the code) a real suggestion but a simple help text because you don't have the span for the `Cargo.toml`

Comment (estebank) [reply tests/ui/check-cfg/cargo-build-script.stderr]: Yeah, I noticed this too as a potential issue. Just for now, I have reverted this change for the ascii art output, but enabling it again sometime in the future would be as simple as changing a single line.

File: compiler/rustc_errors/src/emitter.rs
Comment (estebank) [reply compiler/rustc_errors/src/emitter.rs]: Reverted in a subsequent commit.

File: tests/ui/parser/recover/recover-labeled-non-block-expr.stderr
Comment (estebank) [inline tests/ui/parser/recover/recover-labeled-non-block-expr.stderr:57]: Filed (almost trivial) https://github.com/rust-lang/rust/pull/126723 for these change, that I can rebase on top of once it gets merged.

Comment (estebank) [reply tests/ui/parser/recover/recover-labeled-non-block-expr.stderr]: That's landed now.

File: tests/ui/diagnostic-width/E0271.rs
Comment (fmease) [inline tests/ui/diagnostic-width/E0271.rs:20]: Oh, so `compiletest` can't handle `human-unicode` yet?

**Review by fmease — COMMENTED** [review]
cursory first review comments for the 1st commit

**Comment by fmease** [comment]
Btw, have you checked the perf locally? I presume this change leads to a slight regression but I don't think we can't check this with `rustc-perf` as its benchmark suite is `{check,build,run}-pass` only iirc.

File: compiler/rustc_parse/src/parser/tests.rs
Comment (estebank) [reply compiler/rustc_parse/src/parser/tests.rs:390]: The thin vertical line was on purpose (it looked better to my eyes after a few tries), but is trivially changed. (If we ever support theming, we'll want to make the distinction anyways, even if by default we print it bold throughout.)

File: compiler/rustc_parse/src/parser/tests.rs
Comment (estebank) [reply compiler/rustc_parse/src/parser/tests.rs:337]: Yeah, I had a really *long* conversation and search for an appropriate thing to put there. There's also `╪` which would touch the rest of the line and somewhat work, but nothing like what you drew that would also align with the drawing chars. Another option was the vertical ellipsis `⋮`, but at least on the platforms I had access to absolutely did not look correct (as it didn't come close to aligning).

File: tests/ui/diagnostic-width/E0271.rs
Comment (estebank) [reply tests/ui/diagnostic-width/E0271.rs:20]: Kind of: when modifying error-format we're overriding the implicit `--error-format=json` that compiletest uses. This change predates me handling the forwarding of the underlying rendered field correctly, so we can change this test to be `--error-format=json --json=diagnostic-unicode` and it should Just Work️™️

File: compiler/rustc_errors/src/emitter.rs
Comment (estebank) [reply compiler/rustc_errors/src/emitter.rs:45]: Good callout

File: compiler/rustc_errors/src/emitter.rs
Comment (estebank) [reply compiler/rustc_errors/src/emitter.rs:710]: If that's the case, then we have the same bug elsewhere :)

My logic was that if `unicode_width` was returning nothing for the width of a char, making it take space was a reasonable fallback. I think control characters already return zero, but we can check.

File: compiler/rustc_errors/src/emitter.rs
Comment (estebank) [reply compiler/rustc_errors/src/emitter.rs:1613]: I think after I added this comment I did find where the output was being affected (it was a corner-case IIRC) but failed to extend the comment.

File: compiler/rustc_errors/src/emitter.rs
Comment (estebank) [reply compiler/rustc_errors/src/emitter.rs:2251]: Thanks, I thought I'd handled all of the theming cases already. Yes, this needs to take the string from the theme and the length is variable.

File: compiler/rustc_errors/src/emitter.rs
Comment (estebank) [reply compiler/rustc_errors/src/emitter.rs:2811]: Left over code from when I was figuring out the variables a theme would need.

File: src/librustdoc/core.rs
Comment (estebank) [reply src/librustdoc/core.rs:143]: Good catch.

File: src/librustdoc/doctest.rs
Comment (estebank) [reply src/librustdoc/doctest.rs:435]: Oversight.

File: tests/ui/codemap_tests/huge_multispan_highlight.rs
Comment (estebank) [reply tests/ui/codemap_tests/huge_multispan_highlight.rs:3]: Do `compile-flags` compose like that? I didn't try.

File: tests/ui/codemap_tests/huge_multispan_highlight.unicode.svg
Comment (estebank) [reply tests/ui/codemap_tests/huge_multispan_highlight.unicode.svg:1]: I'm not sure I see what you're seeing. The span (and "underline") do start on the `"`, but because we're using box drawing chars the underline starts "at the middle" of the char, which isn't my preference, but unicode box drawing chars don't give me the flexibility to draw in a nice way that it will cover the start of the char.

File: tests/ui/parser/bad-char-literals.stderr
Comment (estebank) [reply tests/ui/parser/bad-char-literals.stderr:13]: Yeah. This is a pre-existing issue that I'm gonna have to figure out. The current change is "incidental", note that the previous output was also wrong. This is because there's a "zero-width but 1-line-tall char" involved, and the label layout algo doesn't account for that. I'd want to handle that in a separate PR (and consider the new output to be a slight improvement).

File: tests/ui/codemap_tests/huge_multispan_highlight.rs
Comment (estebank) [reply tests/ui/codemap_tests/huge_multispan_highlight.rs]: Done

**Comment by estebank** [comment]
> Btw, have you checked the perf locally?

I haven't, not really. I'll do a smoke test, but I am not sure where you are concerned a regression might be coming from. Just from the reworking, from the additional logic to check what to print, or from the additional bytes that might be taken from non-ascii unicode chars being printed to the stderr bytestream?

File: tests/ui/parser/bad-char-literals.stderr
Comment (estebank) [reply tests/ui/parser/bad-char-literals.stderr]: Fixed

File: compiler/rustc_errors/src/emitter.rs
Comment (estebank) [reply compiler/rustc_errors/src/emitter.rs:1634]: Added a comment documenting when that vertical line is used.

**Comment by apiraino** [comment]
@estebank do we know what the situation is, about Unicode support for all the platforms that rustc supports?
I'm a fan of UTF-8 but I wonder if this would break diags on some Tier 2/3 we don't actively test.

Can you say more about the reasoning behind this change? Why do we want to do this? thanks

**Comment by estebank** [comment]
@apiraino 

> do we know what the situation is, about Unicode support (on terminals) for all the platforms that rustc supports? I'm a fan of UTF-8 but I wonder if this would break diags on some Tier 2/3 we don't actively test. Or on some terminals.

It's spotty at best. I've tested a few different terminals in different platforms and it is honestly quite a terrible status-quo. Mac terminals seem to all support this output correctly. Windows terminals it's hit or miss. Linux is almost always terrible (unless you have a good combination of terminal + fonts, Konsole for example looks ok).

The idea for the output is to be configurable in `~/.cargo/config.toml`, maybe with different default per-platform, but to make this something that individual users can set and forget.

> Can you say more about the reasoning behind this change? Why do we want to do this? thanks

There are a couple of things behind this:

 - I keep finding people that have small misunderstandings with the ASCII output: it is not quite as readable as I'd want. I find that the new output is way more readable, particularly for overlapping multiline spans
 - I find the unicode output more pleasing to the eyes, and communicates the same amount of information with slightly more compact output in some cases
 - I wanted to experiment with what was *possible* while still being constrained by the terminal
 - I wanted to identify what the base components would be if we wanted to ever support theming of the output: I implemented this in such a way that an alternative way of rendering this output is actually possible
 - we've got a goal to migrate to `annotate-snippets`, I prototyped this here but I want for all this code to eventually just live there
 - I want `annotate-snippets` to have this rendering option as a carrot for people to try out the new rendering, I know people would jump at the option
 - I wanted to exercise the `cargo`/`rustc` communication in order to enable this. I encountered a bunch of places where we were failing to properly propagate the information through for anything other than `short` error messages (particularly the interaction with `json` output)
 - "I just think it's neat"

If this lands before we move to annotate-snippets, we'd change the setting for this to actually affect annotate-snippets, so anyone that starts opting-into unicode output will be "enrolled" in the testing of annotate-snippets when we do that move.

File: compiler/rustc_errors/src/emitter.rs
Comment (estebank) [reply compiler/rustc_errors/src/emitter.rs:45]: Added a commit addressing this and separately filed #128806.

File: compiler/rustc_errors/src/emitter.rs
Comment (estebank) [reply compiler/rustc_errors/src/emitter.rs]: This is now better handled in master.

File: compiler/rustc_errors/src/emitter.rs
Comment (estebank) [reply compiler/rustc_errors/src/emitter.rs:45]: done and rebased

**Comment by apiraino** [comment]
Discussed in t-compiler triage [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-09-05/near/467871669)

Green light :vertical_traffic_light: since it will be nightly-only and opt-in anyway

@rustbot label -I-compiler-nominated

File: compiler/rustc_errors/src/emitter.rs
Comment (fmease) [inline compiler/rustc_errors/src/emitter.rs:1085]:
```suggestion
                    buffer.putc(
                        p,
                        (code_offset + annotation.start_col.display).saturating_sub(left),
                        match annotation.annotation_style {
                            AnnotationType::MultilineLine(_) => underline.multiline_vertical,
                            _ => underline.vertical_text_line,
                        },
                        underline.style,
                    );
```

File: compiler/rustc_errors/src/emitter.rs
Comment (fmease) [inline compiler/rustc_errors/src/emitter.rs:1104]: similarly

File: compiler/rustc_errors/src/emitter.rs
Comment (fmease) [inline compiler/rustc_errors/src/emitter.rs:1158]: Could you maybe leave a comment why this is necessary?

File: compiler/rustc_errors/src/emitter.rs
Comment (fmease) [inline compiler/rustc_errors/src/emitter.rs:2252]: Side note: For `annotate-snippet`, it would be nice if symbols like `+` that currently don't vary by `OutputTheme` get moved to the "palette".

**Review by fmease — APPROVED** [review]
r=me after rebase. sorry for the abysmal delay. I have some small nits but they shouldn't really block this PR. esp. since `annotate-snippet` will likely remedy them

**Comment by lqd** [comment]
No worries, `deep-vector` is being noisy.

image

Probably the small period of calm below has lowered the significance threshold, making its usual (seemingly bad) variance trigger as regressions. This should """fix""" itself soon.

image


@rustbot label: +perf-regression-triaged

**Pull request merged**
